Add article about smart contract source code verification. (#2522)
stellar/stellar-docs: docs/build/smart-contracts/getting-started/creating-verifiable-contracts.mdx

:::

Store the fully-qualified, digest-pinned image in a variable so we can reuse it consistently. The following example uses the Docker image for Stellar CLI 27.0.0 for arm64.

<Tabs groupId="platform" defaultValue={getPlatform()}>

<TabItem value="unix" label="macOS/Linux">

```sh
IMAGE="docker.io/stellar/stellar-cli@sha256:c1297d0c2c6790dda6afaa3edd39a959ec12edd6ebe30282dd1d7a663e7c4109"
```

</TabItem>

<TabItem value="windows" label="Windows (PowerShell)">

```powershell
$IMAGE = "docker.io/stellar/stellar-cli@sha256:c1297d0c2c6790dda6afaa3edd39a959ec12edd6ebe30282dd1d7a663e7c4109"
```

</TabItem>

</Tabs>

To keep an in-source toolchain selector (such as a `rust-toolchain.toml` in your project) from silently swapping the toolchain mid-build, export `RUSTUP_TOOLCHAIN` to the image's default before building. We'll pass this variable into the container in [Step 3](#step-3-build-with-embedded-metadata).

<Tabs groupId="platform" defaultValue={getPlatform()}>

<TabItem value="unix" label="macOS/Linux">

```sh

> RUSTUP_TOOLCHAIN=$(docker run --rm --entrypoint rustup "$IMAGE" default | cut -d' ' -f1)

Author: Good catch — fixed in 8e230dd5 by adding `export` so `docker run -e RUSTUP_TOOLCHAIN` forwards the value.

```

</TabItem>

<TabItem value="windows" label="Windows (PowerShell)">

```powershell

> $RUSTUP_TOOLCHAIN = (docker run --rm --entrypoint rustup $IMAGE default).Split(" ")[0]

Author: Fixed in 8e230dd5 — now uses `$env:RUSTUP_TOOLCHAIN` so the container actually receives it.

```

</TabItem>

</Tabs>

## Step 2: Create and hash the source archive

Verifiers need the exact source that produced the contract. Package it into a deterministic archive whose files live within a **single top-level directory**, so extracting it yields exactly one directory holding the source tree.

`git archive` is a convenient way to produce such an archive from a specific commit:

```sh

> git archive --format=tar.gz --prefix=source/ HEAD > hello-world-v1.0.0.tar.gz

Author: Addressed in f49292b4. Switched the redirect to `git archive --output=...` so it doesn't rely on shell redirection (which corrupts binary output on Windows PowerShell 5.1) — no piping needed. On portability/determinism: the archive doesn't need to be reproducibly *re-generated*. Per SEP-58 the verifier downloads the exact archive from `source_uri` and checks it against `source_sha256`; they never rebuild it. So the model is generate-once-and-store. I added prose making that explicit, plus a caution to host the archive yourself rather than relying on auto-generated tarball links whose checksums aren't guaranteed stable.

```

Now compute the SHA-256 of the archive's bytes—this value goes into `source_sha256`:

<Tabs groupId="platform" defaultValue={getPlatform()}>

<TabItem value="unix" label="macOS/Linux">

```sh

> sha256sum hello-world-v1.0.0.tar.gz

Author: Fixed in 8e230dd5 — switched to `shasum -a 256`, which is available on both macOS and Linux.

---
sidebar_position: 60
sidebar_label: "6. Creating Verifiable Contracts"
title: "Creating Verifiable Contracts"
description: "Learn how to create verifiable smart contracts on the Stellar blockchain, ensuring transparency and trust in your decentralized applications."
pagination_next: build/apps/dapp-frontend
---

# 6. Creating Verifiable Contracts

import Tabs from "@theme/Tabs";
import TabItem from "@theme/TabItem";
import { getPlatform } from "@site/src/helpers/getPlatform";

> When you deploy a contract, only its compiled Wasm bytecode lives on-chain. Anyone can read that bytecode, but it tells them nothing about _which source code_ produced it. A **verifiable contract** closes that gap: it embeds enough information for an independent third party to take the published source, rebuild it, and confirm that the result is byte-for-byte identical to what's deployed.

Reviewer: "closes that gap" sounds like AI filler.

Suggest rewording:
```suggestion
When a contract is deployed, anyone can read that bytecode, but it tells them nothing about _the original source code_ that produced it. A **verifiable contract** embeds enough information for an independent third party to discover the published source, rebuild it, and confirm that the result is byte-for-byte identical to what's deployed.
```

> This is what [SEP-58](https://github.com/stellar/stellar-protocol/blob/master/ecosystem/sep-0058.md) standardizes. It defines a shared vocabulary for the build environment information needed to **reproduce a contract's Wasm bytes from source**, so that independent tools can interoperate without prescribing a single workflow.

Reviewer: ```suggestion
[SEP-58](https://github.com/stellar/stellar-protocol/blob/master/ecosystem/sep-0058.md) standardizes the build information required to be able to reproduce a contract's Wasm bytes from source and a set of meta fields for when that data is embedded in a contract.
```

<Tabs groupId="platform" defaultValue={getPlatform()}>

<TabItem value="unix" label="macOS/Linux">

```sh
IMAGE="docker.io/stellar/stellar-cli@sha256:c1297d0c2c6790dda6afaa3edd39a959ec12edd6ebe30282dd1d7a663e7c4109"
```

</TabItem>

<TabItem value="windows" label="Windows (PowerShell)">

```powershell
$IMAGE = "docker.io/stellar/stellar-cli@sha256:c1297d0c2c6790dda6afaa3edd39a959ec12edd6ebe30282dd1d7a663e7c4109"
```

</TabItem>

</Tabs>

To keep an in-source toolchain selector (such as a `rust-toolchain.toml` in your project) from silently swapping the toolchain mid-build, export `RUSTUP_TOOLCHAIN` to the image's default before building. We'll pass this variable into the container in [Step 3](#step-3-build-with-embedded-metadata).

<Tabs groupId="platform" defaultValue={getPlatform()}>

<TabItem value="unix" label="macOS/Linux">

```sh
export RUSTUP_TOOLCHAIN=$(docker run --rm --entrypoint rustup "$IMAGE" default | cut -d' ' -f1)
```

</TabItem>

<TabItem value="windows" label="Windows (PowerShell)">

```powershell

> $env:RUSTUP_TOOLCHAIN = (docker run --rm --entrypoint rustup $IMAGE default).Split(" ")[0]

Reviewer: Does this happen inside the image automatically?
